documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+    JoystickState IReadOnlyList<JoystickState>.this[int index] { get; }
+    
+    public <implementation defined> GetEnumerator();
+    IEnumerator<JoystickState> IEnumerable<JoystickState>.GetEnumerator();
+
+    public event Action<ButtonEvent<JoystickState, JoystickButton>>? ButtonDown;
+    public event Action<ButtonEvent<JoystickState, JoystickButton>>? ButtonUp;
+    public event Action<AxisEvent<JoystickState, float>>? AxisMove;
+    public event Action<AxisEvent<JoystickState, Vector2>>? HatMove;
+}
+```
+
+Unlike 1.0 and 2.0, this proposal uses `readonly record struct`s as their only argument for the event action. This allows us to provide more information to the event handlers without breaking in the future. These types are farily simple:
+
+```cs
+public readonly record struct ButtonEvent<TDevice, TButton>(in TDevice DeviceState, TButton Button);
+public readonly record struct ClickEvent<TDevice, TButton>(in TDevice DeviceState, TButton Button, Vector2 Position);
+public readonly record struct AxisEvent<TDevice, TAxis>(in TDevice DeviceState, int Index, TAxis OldValue, TAxis NewValue);
+public readonly record struct TextEvent<TDevice>(in TDevice DeviceState, string? OldText, string? NewText);
+```
+
+This is the part of this proposal that incorporates the ideas in Enhanced Input Events, and is why this proposal supersedes that one.
+
+One final point to note is that throughout the rest of the proposal the following type will be used:
+
+```cs
+public struct InputReadOnlyList<T> : IReadOnlyList<T>
+{
+    public InputReadOnlyList(IReadOnlyList<T> other);
+}
+```
+
+The Silk.NET team wishes to reserve the right to add more constructors to this type as it sees fit.
+
+This exists so that, should the Silk.NET choose to, we can optimize the lookup of elements while ensuring things like indexers are inlined and don't result in a virtual call if our implementation allows us to do so.
+
+**INFORMATIVE TEXT:** For example, for joystick and mouse buttons we could use a fixed-sized bit buffer where each bit represents an individual button: 1 for pressed, 0 for unpressed. But for something like keyboard input where there are a large amount of keys, we can't do that and will likely use `Memory<T>` instead.
+
+# Mouse Input
+
+As discussed earlier, the interface will be very simple.
+
+```cs
+public interface IMouse : IInputDevice
+{
+    MouseState State { get; }
+    ICursorConfiguration Cursor { get; }
+    void SetPosition(Vector2 pos);
+}
+```
+
+`State` is the device state as defined earlier.
+
+`Cursor` contains the cursor configuration. This isn't actually state that the end user can change, and has been made an interface rather than a state struct accordingly.
+
+`SetPosition` allows moving the mouse cursor without the end user physically moving their mouse.
+
+The device state returned by `State` fills out the following structure:
+
+```cs
+public struct MouseState
+{
+    public IMouse Device { get; init; }
```

Comment:
I thought the backend is free to re-use objects?
In that case we can't reference it here cause if a backend really does pooling this reference might point back to a now differently used object.

Reply:
Left intentionally undefined, but we can disallow this if you want?

Reply:
My argument would be for removing the property as pooling is a good idea, and having this backwards reference is sort of unnecessary.
Removing it also creates a better sense of isolation for functions that operate on the state, as without the property they can't just update their state on their own

Reply:
I don't really want isolation though, as that will just annoy users who don't care about immutability or isolation, some users _just_ want to be able to get something that represents a mouse and check whether that has a button pressed, or move a cursor.

Reply:
I think this leads to too many ways to use the API - I think it should be obvious what the "correct" way to use it is.
Right now we have multiple ways to find out about button presses, have multiple ways to modify devices, and it's not even very obvious which object I'm supposed to be passing around. Should my game loop receive an IMouse or a MouseState?
I think having this tiny ease of use features makes the whole API much less obvious to use.

Reply:
fine.

Reply:
Resolved - devices are no longer included in states.